python/hail/expr/functions.py

```diff
@@ -78,9 +78,14 @@ def lit(x, dtype=None):
     -------
     :class:`.Expression`
     """
+    imputed = dtype is None
     if dtype is None:
         dtype = impute_type(x)
-    dtype._typecheck(x)
+    try:
+        dtype._typecheck(x)
+    except TypeError as e:
+        raise TypeError("'lit': object did not match the {} type '{}'"
```

Comment:
If _typecheck fails in the imputed case, isn't there a bug in impute_type?  I think you can write:
```
if dtype:
  try:
    dtype._typecheck(x)
  except TypeError as e:
    raise ...
else:
  dtype = impute_type(x)
...
```

Reply:
Maybe with an assertion in the second case.